Pretty Printing

wolfv opened the pull request:
This PR adds pretty printing, much like NumPy -- aligning floating point numbers on the `.` etc.

E.g.

```
	xt::xarray<double> rn = xt::random::rand<double>({3, 3}, 0, 1000);
	xt::xarray<double> brn = xt::random::rand<double>({3, 3}, 0, 10000);
	xt::xarray<double> z({3, 3}, 0);
	z(1, 1) = 0.1;
	rn(1, 1) = 0;
	// rn(1, 2) = 10e10;
	rn(1, 2) = -10e5;
	xt::xarray<bool> rb = xt::random::randint<int>({3, 3}, 0, 2);
	xt::xarray<uint> ri = xt::random::randint<int>({3, 3}, 0, 50);
	ri(1, 2) = 10e7;
	pretty_print(rn);
	pretty_print(brn);
	pretty_print(rb);
	pretty_print(ri);
	pretty_print(z);
```

```
{{  1.35477004e+02,   8.35008590e+02,   9.68867771e+02},
 {  2.21034043e+02,   0.00000000e+00,  -1.00000000e+06},
 {  1.88381976e+02,   9.92881302e+02,   9.96461326e+02}}
{{ 9676.9494  ,  7258.3896  ,  9811.0969  },
 { 1098.6175  ,  7981.0586  ,  2970.2945  },
 {   47.834844,  1124.6452  ,  6397.6336  }}
{{ true,  true,  true},
 { true,  true,  true},
 {false, false,  true}}
{{       10,        46,        34},
 {       33,        19, 100000000},
 {       37,        37,        23}}
{{ 0. ,  0. ,  0. },
 { 0. ,  0.1,  0. },
 { 0. ,  0. ,  0. }}
```

SylvainCorlay commented:
@wolfv is on 🔥

wolfv commented:
Hehe, yeah, now I have to stop, don't wanna burn out :)

JohanMabille commented:
It looks like the build has timed out (everything was ok except GCC 4.9 build which didn't start), I've restarted it.

JohanMabille commented:
Is there a reason you reverted the changes made to ``operator<<`` and let the ``pretty_print`` uncalled ?

I think ``pretty_print`` should be the default mode for outputting ``xexpression`` objects (see #94), whatever the output stream is. In order to do so: 

- it should take an additional output stream argument, and return it.
- ``operator<<`` should call ``pretty_print``. Maybe a recursion depth limit should be added, just as in the current implementation of ``operator<<``. @SylvainCorlay any thought on this ?
- current precision of the stream should be checked; if it is greater than the default one used by pretty_print, then use it.

I would really like to have this feature in the next release, it's a huge added value !

wolfv commented:
printing complex numbers depends on #97 now :)

JohanMabille commented:
The error on OSX is due to a Travis problem, I'm restarting it.

wolfv commented:
There are 2 behaviors of numpy that are currently not implemented but might be nice to have:

 - Limiting the line length to 75 chars
 - Limiting the amount of output (putting some ...) for arrays with more than 1000 elements i think. Or actually, the heuristic might be different, need to compare source code.

I think both are a bit harder to achieve. Should we postpone them?

wolfv commented:
Ok, I'll have to refactor this function a bit to make it lazy enough to be worthy as part of xtensor :) 
Ie. I don't want to evaluate the entire array/tensor every time for printing, if only a subset is actually given out on screen.

wolfv commented:
- [ ] Large exponent handling (e+100, e-100 ... )
- [ ] inf/nan handling

JohanMabille commented:
I think we can postpone the 2 behaviors you mentioned. The build issue on OsX seems legitimate, however a rebase on master should fix it (I hope !)

wolfv commented:
Okay, I have more or less rewritten it another time. Now everything *should* work (or at least potentially).

Still need to fix the tests ... 

Let me know if you have doubts about this approach!

JohanMabille commented on include/xtensor/xio.hpp:
```diff
@@ -87,280 +87,356 @@ namespace xt
+        return pretty_print(e, out);
+    }
+
+    inline void set_precision(int prec, std::ostream& out = std::cout)
```
I'm not sure we need a wrapper around ``out << std::setprecision(...)`` but maybe you have some use case ?

JohanMabille commented on include/xtensor/xio.hpp:
```diff
@@ -87,280 +87,356 @@ namespace xt
 
     template <class E>
-    inline std::ostream& operator<<(std::ostream& out, const xexpression<E>& e)
+    std::ostream& pretty_print(const xexpression<E>& e, std::ostream& out = std::cout, unsigned int precision = 8)
```
``precision`` argument is never used after. Moreover, I think something like:
```cpp
if(precision > out.precision())
{
    out << std::setprecision(precision);
}
```
should prevent the default argument to override a higher precision set before.

JohanMabille commented:
For me the approach is good and the code is really cool ! Just a few minor remarks in my review, but that's really minor.

The third commit reverts a lot of things done in the second one, I think that would make sens to squash them in a single one, what do you think ?

wolfv commented:
Hey @JohanMabille, I think it's best I squash when everything is done. 

I have added some test, and for the tests I've added some text files to compare against.

Windows doesn't seem to be able to read them ... maybe it's because the relative path is "wrong" (have to use `\` instead of `/`)?

Otherwise, I think I have implemented my wishlist of features now :)

Do you think it's necessary to have a "fast printing" option (basically restoring the old behavior)?

wolfv commented:
Is there some way I can check wether I'm on windows or unix to adjust the file paths? Or should I try a different approach?

SylvainCorlay commented:
There is http://en.cppreference.com/w/cpp/experimental/fs/path/append

Also regarding "fast printing", I am not sure this is really something meant to be fast.

We are going to tackle serialization in a different package. (probably to npy files first).

wolfv commented:
But are you okay with having the expected output in seperate files?

I think on linux it would work, if the command was

`make xtest` 

or the test was executed by `./test/test_xtensor` instead of changing the directory first.

wolfv commented:
another option might be to copy the files into the `test/` directory with CMake?

wolfv commented:
Ok this shall build now :) Let me know if you have an issue with the way I added the tests. Otherwise I'll add some more in time, and then I hope this thing is ready!

wolfv commented:
Ok, so unfortunately the custom formatter functionality was segfaulting with clang and MSVC. Maybe one of you has an idea why. 

Otherwise, I think functionalitywise it looks good. I'll try to document the print_options a bit.

Cheers!

wolfv commented:
I have rebased once more :) I consider this pretty much done, but let me know if you want to change something! 

One thing that could reduce the code size a bit would to have a base class `base_printer` that defines the (often unnecessary) constructor ...  i might add that later.

SylvainCorlay commented:
I think that we all want this in `0.4.0` but I have not taken the time to read this thoroughly.

I trust @JohanMabille on merging this if he wants to do it now otherwise, I am AFK this afternoon, so this will probably go in tomorrow.

JohanMabille commented:
I think we can merge this PR, the code is really good, the feature is really cool and there's a lot of work involved in this patch to have the cleaner code possible.
